### PR TITLE
Fix missing deepspeed distributed call

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Related #9521 

After https://github.com/PyTorchLightning/pytorch-lightning/pull/9044 DeepSpeed was not updated to use the overridden function `init_ddp_connection`. I've refactored the code such that DeepSpeed handles the necessary initialisation.

As noted by Adrian, #9044 was not included in 1.4.x, so this will be a fix for 1.5

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/plugins/test_deepspeed_plugin.py [hunk 1]
```diff
@@ -383,12 +384,15 @@ def on_before_accelerator_backend_setup(self, trainer, pl_module) -> None:
...
+    """Test to ensure that DeepSpeed with multiple GPUs works and deepspeed distributed is initialized
+    correctly."""
     model = BoringModel()
     trainer = Trainer(
         default_root_dir=tmpdir, plugins=[DeepSpeedPlugin(stage=3)], gpus=2, fast_dev_run=True, precision=16
     )
-    trainer.fit(model)
+    with mock.patch("deepspeed.init_distributed", wraps=deepspeed.init_distributed) as mock_deepspeed_distributed:
+        trainer.fit(model)
+    mock_deepspeed_distributed.assert_called_once()
```
Comment (SeanNaren) [inline tests/plugins/test_deepspeed_plugin.py:395]: I added the test inside another test as a separate test would be expensive to run on CI for this one case.

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 2]
```diff
@@ -334,16 +335,28 @@ def _load_config(self, config):
...
+        # set the ranks and devices
+        self.dist.rank = self.global_rank
+        self.dist.device = self.root_device
         if not self._config_initialized:
             self._format_config()
             self._config_initialized = True
 
-    def init_ddp_connection(self, global_rank: Optional[int] = None, world_size: Optional[int] = None) -> None:
+    def init_deepspeed_distributed(self) -> None:
         if platform.system() != "Windows":
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/deepspeed.py:356]: Does DeepSpeed work well with Windows?

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: tests/plugins/test_deepspeed_plugin.py (hunk 1 above)
Comment (awaelchli) [reply tests/plugins/test_deepspeed_plugin.py:395]: I believe that's very reasonable and also a good choice of test case here

File: CHANGELOG.md [hunk 3]
```diff
@@ -384,6 +384,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed error reporting in DDP process reconciliation when processes are launched by an external agent ([#9389](https://github.com/PyTorchLightning/pytorch-lightning/pull/9389))
 
 
+- Fixed missing deepspeed distributed call ([#9540](https://github.com/PyTorchLightning/pytorch-lightning/pull/9540))
```
Comment (awaelchli) [inline CHANGELOG.md:387]: was #9540 included in the release? I can't see that it did.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 4]
```diff
@@ -334,16 +335,28 @@ def _load_config(self, config):
...
 
     def setup_distributed(self):
-        super().setup_distributed()
+        reset_seed()
+
+        # determine which process we are and world size
+        self.set_world_ranks()
+
+        # set warning rank
+        rank_zero_only.rank = self.global_rank
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:344]: isn't this done in `set_world_ranks`?

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 5]
```diff
@@ -334,16 +335,28 @@ def _load_config(self, config):
...
             self._config_initialized = True
 
-    def init_ddp_connection(self, global_rank: Optional[int] = None, world_size: Optional[int] = None) -> None:
+    def init_deepspeed_distributed(self) -> None:
         if platform.system() != "Windows":
             # do not set env variables on windows, allow deepspeed to control setup
-            global_rank = global_rank if global_rank is not None else self.cluster_environment.global_rank()
-            world_size = world_size if world_size is not None else self.cluster_environment.world_size()
+            global_rank = self.global_rank if self.global_rank is not None else self.cluster_environment.global_rank()
+            world_size = self.world_size if self.world_size is not None else self.cluster_environment.world_size()
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:359]: noob q: why can't we just use the cluster environment directly?

probably for @awaelchli

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/deepspeed.py:359]: general answer: for plugins (especially custom ones) that don't define a cluster env you would need a none check here which is what the properties do. One could also define a default cluster env that returns rank 0 and world size 1, we haven't explored that. 

for deepspeed specifically, you can probably assume there always is an environment so None is not allowed.

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 2 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:356]: Good question, I think it does support it however I haven't tested this. It would be nice to find a user that has :)

File: CHANGELOG.md (hunk 3 above)
Comment (SeanNaren) [reply CHANGELOG.md:387]: aha! great point :) seems like this will only affect 1.5, will change the milestone!

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 6]
```diff
@@ -334,33 +335,38 @@ def _load_config(self, config):
...
+
+        # set the ranks and devices
+        self.dist.rank = self.global_rank
+        self.dist.device = self.root_device
         if not self._config_initialized:
             self._format_config()
             self._config_initialized = True
 
-    def init_ddp_connection(self, global_rank: Optional[int] = None, world_size: Optional[int] = None) -> None:
+    def init_deepspeed_distributed(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/deepspeed.py:352]:
```suggestion
    def _init_deepspeed_distributed(self) -> None:
```
In case there's confusion of this being part of the plug-in interface

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
@SeanNaren @tchaton would you be in favor of making deepspeed plug-in inherit from ParallelPlugin directly? Flattening the inheritance structure reduces the risk of bugs like this
